src/sender.ts

```diff
@@ -5,16 +5,303 @@ import BigNumber from 'bignumber.js'
+ *   const { destinationAccount, sharedSecret } = await getAddressAndSecretFromReceiver()
+ *
+ *   const { destinationAmount, data } = await sendRequest(myLedgerPlugin, {
+ *     destinationAccount,
+ *     sharedSecret,
+ *     destinationAmount: '1000',
+ *     data: Buffer.from('hello', 'utf8')
+ *   })
+ *
+ *   console.log(`Sent payment of: 1000. Receiver got: ${result.destinationAmount}`)
+ *   console.log(`Receiver responded: ${data.toString('utf8')}`)
+ *   // Note the data encoding and content is up to the application
+ * ```
+ *
+ * @example <caption>Quote (unfulfillable test payment)</caption>
+ * ```typescript
+ *   import { sendRequest } from 'ilp-protocol-psk2'
+ *
+ *   // These values must be communicated beforehand for the sender to send a payment
+ *   const { destinationAccount, sharedSecret } = await getAddressAndSecretFromReceiver()
+ *
+ *   const { destinationAmount, data } = await sendRequest(myLedgerPlugin, {
+ *     destinationAccount,
+ *     sharedSecret,
+ *     destinationAmount: '1000',
+ *     unfulfillableCondition: 'random'
+ *   })
+ *   const rate = destinationAmount.dividedBy('1000')
+ *
+ *   console.log(`Path exchange rate is: ${rate}`)
+ * ```
+ */
+export async function sendRequest (plugin: PluginV2, params: SendRequestParams): Promise<PskResponse | PskError> {
+  const debug = Debug('ilp-psk2:sendRequest')
+
+  const requestId = (typeof params.requestId === 'number' ? params.requestId : Math.floor(Math.random() * (constants.MAX_UINT32 + 1)))
+  const sourceAmount = new BigNumber(params.sourceAmount)
+
+  // If the minDestinationAmount is provided, use that
+  // Otherwise, set it to 0 unless the unfulfillableCondition is set, in which case we set it to the maximum value
+  // (This ensures that receivers will respond with the amount they received, as long as they check the amount before the condition)
+  let minDestinationAmount: BigNumber
+  if (params.minDestinationAmount !== undefined) {
+    minDestinationAmount = new BigNumber(params.minDestinationAmount)
+  } else if (params.unfulfillableCondition !== undefined) {
+    minDestinationAmount = constants.MAX_UINT64
+  } else {
+    minDestinationAmount = new BigNumber(0)
+  }
+  assert(Number.isInteger(requestId) && requestId <= constants.MAX_UINT32, 'requestId must be a UInt32')
+  assert(sourceAmount.isInteger() && sourceAmount.lessThanOrEqualTo(constants.MAX_UINT64), 'sourceAmount must be a UInt64')
+  assert(minDestinationAmount.isInteger() && minDestinationAmount.lessThanOrEqualTo(constants.MAX_UINT64), 'minDestinationAmount must be a UInt64')
+
+  // TODO enforce data limit
+
+  const pskPacket = encoding.serializePskPacket(params.sharedSecret, {
+    type: encoding.Type.Request,
+    requestId,
+    amount: new BigNumber(params.minDestinationAmount || 0),
+    data: params.data || Buffer.alloc(0)
+  })
+  let fulfillment
+  let executionCondition
+  if (params.unfulfillableCondition === 'zeros') {
```

Comment:
I find this slightly strange. Why not say that you always pass a buffer in here, if we want to give the user control? Generating 32 random bytes is quite easy.